File: Data Structures & Algorithms/count-number-of-islands/submission-2.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        tot_island=0
        tot_row=len(grid)
        tot_col=len(grid[0])

        def is_valid_index(row,col):
            if row>=0 and row<tot_row and col>=0 and col<tot_col:
                return True
            return False
        
        
        def bfs(row,col):
            nonlocal tot_island
            index_q=[]
            index_q.append((row,col))
            tot_island+=1

            while len(index_q)>0:
                r,c=index_q.pop()
                grid[r][c]="2"
                if is_valid_index(r-1,c) and grid[r-1][c]=="1":
                    index_q.append((r-1,c))
                if is_valid_index(r,c-1) and grid[r][c-1]=="1":
                    index_q.append((r,c-1)) 
                if is_valid_index(r+1,c) and grid[r+1][c]=="1":
                    index_q.append((r+1,c))
                if is_valid_index(r,c+1) and grid[r][c+1]=="1":
                    index_q.append((r,c+1))

        for row in range(tot_row):
            for col in range(tot_col):
                if grid[row][col]=="1":
                    bfs(row,col)
        
        return tot_island

File: Data Structures & Algorithms/rotate-matrix/submission-2.cpp
class Solution {
public:
    void rotate(vector<vector<int>>& matrix) {
        int row=matrix.size();
        int col=row;

        for(int i=0;i<row/2;i++) {
            for(int j=0;j<col;j++) {
                int temp=matrix[i][j];
                matrix[i][j]=matrix[row-1-i][j];
                matrix[row-1-i][j]=temp;
            }
        }

        for(int i=0;i<row;i++) {
            for(int j=0;j<i;j++) {
                int temp=matrix[i][j];
                matrix[i][j]=matrix[j][i];
                matrix[j][i]=temp;
            }
        }
    }
};
